internal/provider/framework/intercept.go

```diff
@@ -321,15 +326,122 @@ type interceptedResponse interface {
 		resource.ImportStateResponse
 }
 
+type innerFunc[Request, Response any] func(ctx context.Context, request *Request, response *Response) diag.Diagnostics
+
 // interceptedHandler returns a handler that runs any interceptors.
-func interceptedHandler[Request interceptedRequest, Response interceptedResponse](interceptors []interceptorFunc[Request, Response], f func(context.Context, *Request, *Response) diag.Diagnostics, c *conns.AWSClient) func(context.Context, *Request, *Response) diag.Diagnostics {
-	return func(ctx context.Context, request *Request, response *Response) diag.Diagnostics {
-		var diags diag.Diagnostics
-		// Before interceptors are run first to last.
-		forward := interceptors
+func interceptedHandler[Request interceptedRequest, Response interceptedResponse](interceptors []interceptorFunc[Request, Response], f innerFunc[Request, Response], c awsClient) func(context.Context, *Request, *Response) diag.Diagnostics {
+	return func(ctx context.Context, request *Request, response *Response) (diags diag.Diagnostics) {
+		// We need to stash the diagnostics from the Response to preserve any existing diagnostics because
+		// the `inner` function will actually returns its diagnostics in the Response, but we are collecting them here as well.
+		var stashDiags diag.Diagnostics
+		switch v := any(response).(type) {
```

Comment:
This `switch` statement is needed because `Diagnostics` is a field on `response`, so it cannot be abstracted away by generics. The other option is to make type-specific `interceptedHandler` versions

Reply:
Possible, but not advisable.